src/service.rs

```diff
@@ -1424,14 +1424,17 @@ impl Service {
     /// Get the service config from the system.
     pub fn query_config(&self) -> crate::Result<ServiceConfig> {
         // As per docs, the maximum size of data buffer used by QueryServiceConfigW is 8K
-        let mut data = [0u8; MAX_QUERY_BUFFER_SIZE];
+        let mut buf = unsafe {
+            mem::MaybeUninit::<[mem::MaybeUninit<u8>; MAX_QUERY_BUFFER_SIZE]>::uninit()
+                .assume_init()
```

Comment:
This was safe from the start. `[u8; N]` is a safely initialized array. Nothing around here has potential uninitialized UB.

Reply:
Then I'm not sure why I'm getting a segmentation fault when attempting to run the `service_config` example

```rust
 /// Get the service config from the system.
    pub fn query_config(&self) -> crate::Result<ServiceConfig> {
        // As per docs, the maximum size of data buffer used by QueryServiceConfigW is 8K
        let mut buf = [0u8; MAX_QUERY_BUFFER_SIZE];
        let mut bytes_written: u32 = 0;

        let success = unsafe {
            winsvc::QueryServiceConfigW(
                self.service_handle.raw_handle(),
                buf.as_mut_ptr() as _,
                MAX_QUERY_BUFFER_SIZE as u32,
                &mut bytes_written,
            )
        };

        if success == 0 {
            Err(Error::Winapi(io::Error::last_os_error()))
        } else {
            unsafe {
                // Safety: buf has been initialized
                let raw_config =
                    std::mem::transmute::<_, *const winsvc::QUERY_SERVICE_CONFIGW>(buf.as_ptr());
                ServiceConfig::from_raw(*raw_config)
            }
        }
    }
```

```
cargo run --example service_config
   Compiling windows-service v0.3.0 (C:\Dev\windows-service-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 1.68s
     Running `target\debug\examples\service_config.exe`
error: process didn't exit successfully: `target\debug\examples\service_config.exe` (exit code: 1)
```

Using `MaybeUninit` fixes the issue

```rust
/// Get the service config from the system.
    pub fn query_config(&self) -> crate::Result<ServiceConfig> {
        // As per docs, the maximum size of data buffer used by QueryServiceConfigW is 8K
        let mut buf = mem::MaybeUninit::<[u8; MAX_QUERY_BUFFER_SIZE]>::uninit();
        let mut bytes_written: u32 = 0;

        let success = unsafe {
            winsvc::QueryServiceConfigW(
                self.service_handle.raw_handle(),
                buf.as_mut_ptr() as _,
                MAX_QUERY_BUFFER_SIZE as u32,
                &mut bytes_written,
            )
        };

        if success == 0 {
            Err(Error::Winapi(io::Error::last_os_error()))
        } else {
            unsafe {
                // Safety: buf has been initialized
                let raw_config =
                    std::mem::transmute::<_, *const winsvc::QUERY_SERVICE_CONFIGW>(buf.as_ptr());
                ServiceConfig::from_raw(*raw_config)
            }
        }
    }
```

```
cargo run --example service_config
   Compiling windows-service v0.3.0 (C:\Dev\windows-service-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 0.56s
     Running `target\debug\examples\service_config.exe`
ServiceConfig {
    service_type: SHARE_PROCESS,
    start_type: AutoStart,
    error_control: Normal,
    executable_path: "C:\\WINDOWS\\system32\\lsass.exe",
    load_order_group: Some(
        "MS_WindowsRemoteValidation",
    ),
    tag_id: 0,
    dependencies: [
        Service(
            "LanmanWorkstation",
        ),
    ],
    account_name: Some(
        "LocalSystem",
    ),
    display_name: "Netlogon",
}
```

Any idea ?

Reply:
It looks like when using the `QueryServiceConfigW` API we should be able to get away with initializing `QUERY_SERVICE_CONFIGW` directly as opposed to using the 8k byte array. I think the docs for `QueryServiceConfigW` are copy-and-paste from the docs for `QueryServiceConfig2W` which accepts a flag that controls the kind of information that the generic `lpBuffer` parameter is being filled with. In case of `QueryServiceConfigW` we have a strictly typed buffer.

So either way I think it's reasonable to use `MaybeUninit` here instated of a large byte buffer:

```rs
let mut buf = unsafe {  mem::MaybeUninit::<winsvc::QUERY_SERVICE_CONFIGW>::zeroed() };
```

Obviously the size of the `QUERY_SERVICE_CONFIGW` should be used in place of `MAX_QUERY_BUFFER_SIZE` too.

Docs:

1. https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-queryserviceconfigw
2. https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-queryserviceconfig2w

Reply:
Looking at the example from the documentation, I don't think that's the case.

The MSDN example is first retrieving the number of bytes needed prior to calling `QueryServiceConfig` again:

```cpp
   if( !QueryServiceConfig( 
        schService, 
        NULL, 
        0, 
        &dwBytesNeeded))
    {
        dwError = GetLastError();
        if( ERROR_INSUFFICIENT_BUFFER == dwError )
        {
            cbBufSize = dwBytesNeeded;
            lpsc = (LPQUERY_SERVICE_CONFIG) LocalAlloc(LMEM_FIXED, cbBufSize);
        }
        else
        {
            printf("QueryServiceConfig failed (%d)", dwError);
            goto cleanup; 
        }
    }
```

It looks like the number of bytes needed is not `mem::size_of<QUERY_SERVICE_CONFIGW>()` but really depends on the configuration:

```console
λ cargo run --example service_config -- netlogon                
   Compiling windows-service v0.3.0 (C:\Dev\windows-service-rs) 
    Finished dev [unoptimized + debuginfo] target(s) in 1.39s   
     Running `target\debug\examples\service_config.exe netlogon`
bytes_needed = 354                                              
ServiceConfig {                                                 
    service_type: SHARE_PROCESS,                                
    start_type: AutoStart,                                      
    error_control: Normal,                                      
    executable_path: "C:\\WINDOWS\\system32\\lsass.exe",        
    load_order_group: Some(                                     
        "MS_WindowsRemoteValidation",                           
    ),                                                          
    tag_id: 0,                                                  
    dependencies: [                                             
        Service(                                                
            "LanmanWorkstation",                                
        ),                                                      
    ],                                                          
    account_name: Some(                                         
        "LocalSystem",                                          
    ),                                                          
    display_name: "Netlogon",                                   
}                                                               
```

```console
λ cargo run --example service_config -- SysMain                                               
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s                                 
     Running `target\debug\examples\service_config.exe SysMain`                               
bytes_needed = 394                                                                            
ServiceConfig {                                                                               
    service_type: SHARE_PROCESS,                                                              
    start_type: AutoStart,                                                                    
    error_control: Ignore,                                                                    
    executable_path: "C:\\WINDOWS\\system32\\svchost.exe -k LocalSystemNetworkRestricted -p", 
    load_order_group: Some(                                                                   
        "profsvc_group",                                                                      
    ),                                                                                        
    tag_id: 0,                                                                                
    dependencies: [                                                                           
        Service(                                                                              
            "rpcss",                                                                          
        ),                                                                                    
        Service(                                                                              
            "fileinfo",                                                                       
        ),                                                                                    
    ],                                                                                        
    account_name: Some(                                                                       
        "LocalSystem",                                                                        
    ),                                                                                        
    display_name: "SysMain",                                                                  
}                                                                                             
```

So I think I would keep the 8k byte array approach.

Reply:
Alright. That's possible. In the end they might be returning something that looks like `QUERY_SERVICE_CONFIGW` but is actually a different possibly private struct that just happens to share the same fields.